### PR TITLE
SQL Datasources: Move database setting to jsonData

### PR DESCRIPTION
**Pull request by zoltanbedi**
This PR migrates the database field for SQL datasources under jsonData. This was necessary as we used these fields before but didn't get back values as part of the grafanaBootConfig because of this https://github.com/grafana/grafana/blob/main/pkg/api/frontendsettings.go#L317. After talking with @marefr, he suggested that database field shouldn't be used anymore and should be under jsonData instead.

This PR is based on https://github.com/grafana/grafana/pull/58647. That needs to be merged first.

### CODE REVIEW

File: pkg/tsdb/mssql/mssql.go [hunk 1]
```diff
@@ -65,11 +65,17 @@ func newInstanceSettings(cfg *setting.Cfg) datasource.InstanceFactoryFunc {
 		if err != nil {
 			return nil, fmt.Errorf("error reading settings: %w", err)
 		}
+
+		database := settings.Database
```
Comment (marefr) [inline pkg/tsdb/mssql/mssql.go:69]: Maybe the opposite, check the jsondata first and then fallback to legacy field?

File: pkg/tsdb/mysql/mysql.go [hunk 2]
```diff
@@ -60,11 +60,17 @@ func newInstanceSettings(cfg *setting.Cfg, httpClientProvider httpclient.Provide
 		if err != nil {
 			return nil, fmt.Errorf("error reading settings: %w", err)
 		}
+
+		database := settings.Database
```
Comment (marefr) [inline pkg/tsdb/mysql/mysql.go:64]: Same as above

File: pkg/tsdb/postgres/postgres.go [hunk 3]
```diff
@@ -66,11 +66,17 @@ func (s *Service) newInstanceSettings(cfg *setting.Cfg) datasource.InstanceFacto
 		if err != nil {
 			return nil, fmt.Errorf("error reading settings: %w", err)
 		}
+
+		database := settings.Database
```
Comment (marefr) [inline pkg/tsdb/postgres/postgres.go:70]: Same as above

**Review by marefr — COMMENTED** [review]
Looks good. Added couple minor things. And as I understand any other datasource such as elasticsearch and influx would continue to work?

What about provisioning. Might want to migrate there as well or log a depreciation warning perhaps?

**Comment by zoltanbedi** [comment]
Thanks for the review @marefr!

> And as I understand any other datasource such as elasticsearch and influx would continue to work?

Yes this is right.

> What about provisioning. Might want to migrate there as well or log a depreciation warning perhaps?

Can you please help me where should I do that?

File: pkg/tsdb/mssql/mssql.go (hunk 1 above)
Comment (zoltanbedi) [reply pkg/tsdb/mssql/mssql.go]: Changed it.

**Comment by marefr** [comment]
> Can you please help me where should I do that?

@zoltanbedi after thinking about it you probably want to do this when all affected datasources supports `jsonData.database`. And since you have handled this for SQL datasources to fallback to old database field it should be fine from a provisioning perspective as well.

In the future could imagine checking if the database field is set and either log a deprecation warning or clear it and populate `jsonData.database` instead. Applicable for add/update datasource in service implementation since both HTTP API and provisioning are using these:
https://github.com/grafana/grafana/blob/f9ef0eb414eab92f3f58f46632c437634a2e5e57/pkg/services/datasources/service/datasource.go#L170
https://github.com/grafana/grafana/blob/f9ef0eb414eab92f3f58f46632c437634a2e5e57/pkg/services/datasources/service/datasource.go#L227

**Review by marefr — APPROVED** [review]
LGTM

**Review by codeincarnate — APPROVED** [review]
Tested out with MySql and Postgres data sources it and appears to be working 😄.  Would be cool to have someone else from @grafana/grafana-bi-squad check out MSSQL if they have the time as I don't currently have a data source setup. 

A side comment: Is there any consequence if someone doesn't update their datasource manually? I'm thinking not but not sure if others think this might be an issue.

**Review by mdvictor — APPROVED** [review]
LGTM. Tested for mssql ds successfully!

**Comment by hterik** [comment]
This need to be documented as breaking change for 9.4. Either on https://grafana.com/docs/grafana/latest/upgrade-guide/upgrade-v9.4/ or https://grafana.com/docs/grafana/latest/upgrade-guide/upgrade-v9.4/

**Comment by zoltanbedi** [comment]
Hey @hterik I agree, this should be included in the what's new doc however it is not a breaking change as the database field is still supported. Am I missing something? Can you please share your use case where this breaks?

**Comment by hterik** [comment]
@zoltanbedi  The old format is not working any more, when provisioning the datasource via `datasources.yaml`. Instead the database field in UI is presented as blank and all dashboards using the datasource fail with errors such as **pq: Invalid connection parameters specified. Check connection parameters, and try again.**

**Comment by zoltanbedi** [comment]
@hterik can you please show your yaml file? I've just tried it out with this and I was able to see the database field in the config page and also worked.

```yaml
apiVersion: 1

datasources:
  - name: Postgres provisioning
    type: postgres
    url: localhost:5432
    user: grafana
    database: grafana
    secureJsonData:
      password: 'password'
    jsonData:
      sslmode: 'disable'
      maxOpenConns: 0
      maxIdleConns: 2
      connMaxLifetime: 14400
      postgresVersion: 903
      timescaledb: false
```

**Comment by zoltanbedi** [comment]
You are right @hterik. I was able to reproduce this with a clean grafana. This was a bug in our end but the fix was not backported to 9.4.7 it will be included in 9.4.8 and already in 9.5.0. Thanks for letting me know. I advise to use the new config option but the old should still work in 9.4.8 and 9.5.0. https://github.com/grafana/grafana/pull/62790

**Pull request merged**